File: web_version/pmugg_files/shape/vertexGroup.js
// A vertex.
ms.vertexGroup = function(x, y, vertex) {
	this.position = new ms.vec2(x, y);
	this.vertices = [vertex];
	this.selected = false;
	this.visible = true;
	this.hovered = false;
	this.decoration = null;

	// Just for debugging.
	this.id = ms.vertexGroup.counter++;
};

ms.vertexGroup.prototype.isControl = function() {
	return false;
};

ms.vertexGroup.prototype.isSnappable = function() {
	return true;
};

ms.vertexGroup.prototype.getDecoration = function() {
	return this.decoration;
};

ms.vertexGroup.prototype.setDecoration = function(decoration) {
	this.decoration = decoration;
};

ms.vertexGroup.counter = 0;

ms.vertexGroup.prototype.merge = function(groupB) {
	this.vertices = this.vertices.concat(groupB.vertices);
	var self = this;
	groupB.vertices.forEach(function(vertex) {
		vertex.setGroup(self);
	});
	groupB.vertices = [];
	this.visible = true;
	this.selected = false;
	var decorationB = groupB.getDecoration();
	var imageB = decorationB && decorationB.get('Image');
	if (imageB && imageB != 'None') {
		this.decoration = groupB.getDecoration();
	}
};

ms.vertexGroup.create = function(p) {
	var result = new ms.vertexGroup(0, 0);
	result.position = p.getPosition();
	return result;
};

ms.vertexGroup.prototype.makeUnique = function(vertex) {
	if (this.vertices.length > 1) {
		this.removeVertex(vertex);
		vertex.setGroup(new ms.vertexGroup(this.position.x, this.position.y, vertex));
	}
};

ms.vertexGroup.prototype.removeVertex = function(vertex) {
	ms.remove(vertex, this.vertices);
};

ms.vertexGroup.prototype.getVertex = function() {
	if (this.vertices.length != 1) {
		alert('Wrong number of vertices in getVertex.');
	}
	return this.vertices[0];
};

ms.vertexGroup.fromPosition = function(position) {
	var result = new ms.vertexGroup(0, 0);
	result.position = position;
	return result;
};

ms.vertexGroup.prototype.getPosition = function() {
	return this.position;
};

ms.vertexGroup.prototype.setPosition = function(position) {
	this.position.set(position);
};

ms.vertexGroup.prototype.getSelectedPositions = function() {
	return [this.position];
};

ms.vertexGroup.prototype.getVertices = function () {
	return [this];
};

ms.vertexGroup.prototype.getSelectedVertices = function () {
	return [this];
};

ms.vertexGroup.prototype.getEdges = function() {
	var edges = [];
	for (var i = 0; i < this.vertices.length; i++) {
		edges = edges.concat(this.vertices[i].getEdges());
	}
	return edges;
};

ms.vertexGroup.prototype.getEndpoints = function() {
	var endpoints = [];
	for (var i = 0; i < this.vertices.length; i++) {
		endpoints = endpoints.concat(this.vertices[i].getEndpoints());
	}
	return endpoints;
};

ms.vertexGroup.prototype.getNext = function(endpoint) {
	var endpoints = this.getEndpoints();
	endpoints.sort(function(a, b) {
		return a.tangent() - b.tangent();
	});
	var index = endpoints.indexOf(endpoint);
	if (index == -1) {
		window.console.log('Endpoint not found');
		return null;
	}
	index = (index + 1) % endpoints.length;
	return endpoints[index];
};

ms.vertexGroup.prototype.isLarge =  function () {
	return true;
};

ms.vertexGroup.prototype.getColor =  function () {
	if (this.selected) {
		return ms.shapeView.SELECTED_COLOR;
	} else {
		if (this.hovered) {
			return ms.shapeView.HOVERED_COLOR;
		} else {
			return ms.shapeView.DEFAULT_COLOR;
		}
	}
};

ms.vertexGroup.prototype.getX = function() {
	return this.position.x;
};

ms.vertexGroup.prototype.getY = function() {
	return this.position.y;
};

ms.vertexGroup.prototype.distance = function(p) {
	return this.position.distance(p.getPosition());
};

ms.vertexGroup.prototype.sortEndpoints = function() {
	for (var j = 0; j < this.vertices.length; j++) {
		this.vertices[j].sortEndpoints();
	}
};

ms.vertexGroup.prototype.getSelected = function() {
	return this.selected;
};

ms.vertexGroup.prototype.hover = function() {
	var changed = !this.hovered;
	this.hovered = true;
	return changed;
};

ms.vertexGroup.prototype.unhover = function() {
	var changed = this.hovered;
	this.hovered = false;
	return changed;
};

ms.vertexGroup.prototype.select = function() {
	this.selected = true;
	for (var i = 0; i < this.vertices.length; i++) {
		this.vertices[i].select();
	}
};

ms.vertexGroup.prototype.deselect = function() {
	this.selected = false;
};

ms.vertexGroup.prototype.getVisible = function() {
	return this.visible;
};

ms.vertexGroup.prototype.copy = function() {
	var v = new ms.exampleVertex(this.position.x, this.position.y);
	return v.getGroup();
};

ms.vertexGroup.prototype.directMove = function(dx, dy) {
	this.position.x += dx;
	this.position.y += dy;
};

// TODO: Maybe use the direct move function more that doesn't search for
// hovered control points. This could speed up the exampleRepeater.
ms.vertexGroup.prototype.move = function(dx, dy) {
	if (this.selected) {
		this.position.x += dx;
		this.position.y += dy;
		for (var i = 0; i < this.vertices.length; i++) {
			this.vertices[i].move(dx, dy);
		}
	}
};

// Returns true is the position is close. This is used for selection.
ms.vertexGroup.prototype.isClose = function(query, scale) {
	var position = this.getPosition();
	return (position && query.distance2(position) < ms.shapeSet.NEAR_RADIUS2 / (scale * scale));
};

ms.vertexGroup.prototype.selectType = function() {
	return ms.shapeMaker.SelectableTypes.VERTEX;
};

ms.vertexGroup.prototype.getShape = function() {
	return this.vertices[0].endpoints[0].edge.shape;
};

ms.vertexGroup.prototype.draw = function(view, selected, secondPass) {
	view.drawVertex(this, selected, secondPass);
};

File: cpp_version/pmugg/shape/vertex_group.cpp
#include "vertex_group.h"
#include "vertex.h"
#include "edge.h"
#include "endpoint.h"
#include "decoration.h"
#include "view.h"
#include "shape.h"
#include "util.h"
#include <algorithm>

namespace ms {

int VertexGroup::nextId = 0;

VertexGroup::VertexGroup(float x, float y, Vertex* vertex)
    : position(x, y)
    , decoration(nullptr)
    , selected(false)
    , visible(true)
    , hovered(false)
    , id(nextId++) {
    if (vertex) {
        vertices.push_back(vertex);
    }
}

void VertexGroup::removeVertex(Vertex* vertex) {
    auto it = std::find(vertices.begin(), vertices.end(), vertex);
    if (it != vertices.end()) {
        vertices.erase(it);
    }
}

Vertex* VertexGroup::getVertex() const {
    if (vertices.size() != 1) {
        throw std::runtime_error("Wrong number of vertices in getVertex");
    }
    return vertices[0];
}

void VertexGroup::makeUnique(Vertex* vertex) {
    if (vertices.size() > 1) {
        removeVertex(vertex);
        vertex->setGroup(new VertexGroup(position.getX(), position.getY(), vertex));
    }
}

void VertexGroup::merge(VertexGroup* groupB) {
    vertices.insert(vertices.end(), groupB->vertices.begin(), groupB->vertices.end());
    
    for (auto* vertex : groupB->vertices) {
        vertex->setGroup(this);
    }
    
    groupB->vertices.clear();
    visible = true;
    selected = false;

    auto* decorationB = groupB->getDecoration();
    if (decorationB) {
        auto imageB = decorationB->get("Image");
        if (imageB && imageB != "None") {
            decoration = decorationB;
        }
    }
}

std::vector<Vec2> VertexGroup::getSelectedPositions() const {
    return {position};
}

std::vector<VertexGroup*> VertexGroup::getVertices() const {
    return {const_cast<VertexGroup*>(this)};
}

std::vector<VertexGroup*> VertexGroup::getSelectedVertices() const {
    return {const_cast<VertexGroup*>(this)};
}

std::vector<Edge*> VertexGroup::getEdges() const {
    std::vector<Edge*> edges;
    for (auto* vertex : vertices) {
        auto vertexEdges = vertex->getEdges();
        edges.insert(edges.end(), vertexEdges.begin(), vertexEdges.end());
    }
    return edges;
}

std::vector<Endpoint*> VertexGroup::getEndpoints() const {
    std::vector<Endpoint*> endpoints;
    for (auto* vertex : vertices) {
        auto vertexEndpoints = vertex->getEndpoints();
        endpoints.insert(endpoints.end(), vertexEndpoints.begin(), vertexEndpoints.end());
    }
    return endpoints;
}

Endpoint* VertexGroup::getNext(Endpoint* endpoint) const {
    auto endpoints = getEndpoints();
    std::sort(endpoints.begin(), endpoints.end(), 
        [](Endpoint* a, Endpoint* b) { return a->tangent() < b->tangent(); });
    
    auto it = std::find(endpoints.begin(), endpoints.end(), endpoint);
    if (it == endpoints.end()) {
        std::cerr << "Endpoint not found" << std::endl;
        return nullptr;
    }
    
    return endpoints[(it - endpoints.begin() + 1) % endpoints.size()];
}

void VertexGroup::directMove(float dx, float dy) {
    position.move(dx, dy);
}

void VertexGroup::move(float dx, float dy) {
    if (selected) {
        position.move(dx, dy);
        for (auto* vertex : vertices) {
            vertex->move(dx, dy);
        }
    }
}

float VertexGroup::distance(const Vec2& p) const {
    return position.distance(p);
}

bool VertexGroup::isClose(const Vec2& query, float scale) const {
    return query.distance2(position) < Shape::NEAR_RADIUS2 / (scale * scale);
}

void VertexGroup::select() {
    selected = true;
    for (auto* vertex : vertices) {
        vertex->select();
    }
}

void VertexGroup::deselect() {
    selected = false;
}

bool VertexGroup::hover() {
    bool changed = !hovered;
    hovered = true;
    return changed;
}

bool VertexGroup::unhover() {
    bool changed = hovered;
    hovered = false;
    return changed;
}

int VertexGroup::selectType() const {
    return Shape::SelectableTypes::VERTEX;
}

void VertexGroup::draw(View* view, bool selected, bool secondPass) {
    view->drawVertex(this, selected, secondPass);
}

std::string VertexGroup::getColor() const {
    if (selected) {
        return Shape::SELECTED_COLOR;
    }
    return hovered ? Shape::HOVERED_COLOR : Shape::DEFAULT_COLOR;
}

VertexGroup* VertexGroup::create(const Vec2& p) {
    auto* result = new VertexGroup(0, 0);
    result->position = p;
    return result;
}

VertexGroup* VertexGroup::fromPosition(const Vec2& position) {
    auto* result = new VertexGroup(0, 0);
    result->position = position;
    return result;
}

} // namespace ms 